libs-sol/token-locker/contracts/TokenLocker.sol

```diff
@@ -1,16 +1,116 @@
 pragma solidity ^0.5.0;
 import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
 import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+import "../../nearprover/contracts/INearProver.sol";
+import "../../nearprover/contracts/ProofDecoder.sol";
+import "../../nearbridge/contracts/NearDecoder.sol";
+import "../../nearbridge/contracts/Borsh.sol";
+
 contract TokenLocker {
     using SafeERC20 for IERC20;
+    using Borsh for Borsh.Data;
+    using ProofDecoder for Borsh.Data;
+    using NearDecoder for Borsh.Data;
+
+    IERC20 public ethToken_;
+    bytes public nearToken_;
+    INearProver public prover_;
+
+    mapping(bytes32 => bool) public usedEvents_;
+
     event Locked(
         address indexed token,
         address indexed sender,
         uint256 amount,
         string accountId
     );
-    function lockToken(IERC20 token, uint256 amount, string memory accountId) public {
-        token.safeTransferFrom(msg.sender, address(this), amount);
-        emit Locked(address(token), msg.sender, amount, accountId);
+
+    event Unlocked(
+        uint128 amount,
+        address recipient
+    );
+
+    // Function output from burning fungible token on Near side.
+    struct BurnResult {
+        uint128 amount;
+        address recipient;
+    }
+
+    function _decodeBurnResult(bytes memory data) internal pure returns(BurnResult memory result) {
+        Borsh.Data memory borshData = Borsh.from(data);
+        result.amount = borshData.decodeU128();
+        bytes20 recipient = borshData.decodeBytes20();
+        result.recipient = address(uint160(recipient));
+    }
+
+    function lockToken(uint256 amount, string memory accountId) public {
+        ethToken_.safeTransferFrom(msg.sender, address(this), amount);
+        emit Locked(address(ethToken_), msg.sender, amount, accountId);
+    }
+
+    function uintToString(uint256 v) internal pure returns (string memory str) {
+        uint maxlength = 100;
+        bytes memory reversed = new bytes(maxlength);
+        uint i = 0;
+        while (v != 0) {
+            uint remainder = v % 10;
+            v = v / 10;
+            reversed[i++] = bytes1(uint8(48 + remainder));
+        }
+        bytes memory s = new bytes(i + 1);
+        for (uint j = 0; j <= i; j++) {
+            s[j] = reversed[i - j];
+        }
+        str = string(s);
+    }
+
+    function char(byte b) internal pure returns (byte c) {
+        if (uint8(b) < uint8(10)) return byte(uint8(b) + 0x30);
+        else return byte(uint8(b) + 0x57);
+    }
+
+    function bytes32string(bytes32 b32) internal pure returns (string memory out) {
```

Comment:
This, char and uintToString seems to be reused in a few places, move them to NearDecoder or some util.sol?

Reply:
That was debugging code. Removed.